scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/MetadataJsonValueValidator.scala

```diff
@@ -0,0 +1,392 @@
+      // don't return error as it will be caught when its f2/f3's turn
+      case Some(parentId) if idsOfParents.contains(parentId) => parentIdNel
+      case Some(parentId) if parentId == originalId =>
+        val parentIdThatReferencesEntry = idsOfParents.last
+        val idsOfParentsWithId = List(parentId) ::: idsOfParents ::: List(parentId)
+        val breadcrumbOfIds = idsOfParentsWithId.mkString(" > ")
+        HierarchyLinkingError(
+          parentIdThatReferencesEntry,
+          s"Circular dependency! A parent entry (id '$parentIdThatReferencesEntry') references this entry as its parentId.\n\n" +
+            s"The breadcrumb trail looks like this $breadcrumbOfIds"
+        ).invalidNel[Value]
+      case Some(parentId) => getCircularDependencyErrors(allFolderTypeEntryIds, originalId, parentId, idsOfParents ::: List(parentId), parentIdNel)
+    }
+  }
+
+  def getIdsOfAllEntries(allEntries: Map[String, List[ValidatedEntry]], metadataFileIds: List[String] = Nil): List[(FieldName, EntryTypeAndParent)] = {
+    def getEntryType(entryTypeAsString: String, potentialParentId: Option[String]): EntryTypeAndParent =
+      entryTypeAsString match {
+        case "File"            => FileEntry(potentialParentId)
+        case "MetadataFile"    => MetadataFileEntry(potentialParentId)
+        case "UnknownFileType" => UnknownFileTypeEntry(potentialParentId)
+        case "Asset"           => AssetEntry(potentialParentId)
+        case "ArchiveFolder"   => ArchiveFolderEntry(potentialParentId)
+        case "ContentFolder"   => ContentFolderEntry(potentialParentId)
+        case "UnknownType"     => UnknownTypeEntry(potentialParentId)
+      }
+
+    allEntries.toList.flatMap { case (entryType, entries) =>
+      entries.flatMap { entry =>
+        val idNel = entry(id)
+        val parentIdNel = entry("parentId")
+        val entryTypeUpdated =
+          if entryType == "File" then
+            val nameNel = entry("name")
+            nameNel match {
+              case Validated.Valid(nameValue) =>
+                val idOfFile = idNel.getOrElse(Str("idFieldHasErrorInIt")).str
+                val nameAsString = nameValue.str
+                if nameAsString.endsWith("-metadata.json") || metadataFileIds.contains(idOfFile) then "MetadataFile" else entryType
+              case invalidNel => "UnknownFileType"
+            }
+          else entryType
+
+        idNel match {
+          case Validated.Valid(idValue) =>
+            val idAsString = idValue.str
+            parentIdNel match {
+              case Validated.Valid(parentIdValue) =>
+                val potentialParentIdAsString = parentIdValue.strOpt
+                val entryValue = getEntryType(entryTypeUpdated, potentialParentIdAsString)
+                List((idAsString, entryValue))
+              case invalidNel => List((idAsString, getEntryType(entryTypeUpdated, Some("parentId undetermined, due to validation error"))))
```

Comment:
I'm not 100% sure I understand your comment, when you say this "Why do you need this?", do you mean why do I add `case invaliNel =>` in general or do you mean, why do I have it in this instance? 

If in general:

In some cases I do it so that I can retain the error message (and so that I don't overwrite the first message because we want that to be fixed first) and in other cases, I do it because a method (like this one) requires one or more fields from the entry and if the field isn't there, it's fine.

If in this instance:

In this particular instance, this line doesn't have any effect on the entry's `parentId` in the Json Map. This `getIdsOfAllEntries` method is different from the other methods as it generates a list of `(id, EntryType(potentialParentId))`, so that a lookups can be done with it later, as you can see from [this line](https://github.com/nationalarchives/dr2-ingest/pull/242/files/004e14dca89cd673d2ef2401d8b4af400bef80ff#diff-a7717b36922cd831025382930e0363332790000ed617c05471e451397ccc5980R113) in the lambda; if I returned Nil, then the list would be missing an entry. `getEntryType` will return a more fine-grained type (in the case of a File), with the `parentId` inside. I added this line so that if the `parentId` is undetermined, I didn't want to put `None` because that is a legitimate value (a top-level folder can have a null/None `parentId`)

Reply:
Ok, that makes sense, leave it as is then.